14409 8 improve

belliottsmith opened the pull request.

Commit:
Replica Layouts
Fix how natural is written
Fix
Fix problems with batch log writes, make sure pending replicas are included
Remove blockFor as it has more complicated semantics
Make sure to collect additional replicas comparing from all, not from natural
Make argument order more consistent

belliottsmith committed:
Require explicit InetAddressAndEndpoint in RangesAtEndpoint constructor
Serialize endpoint separately in StreamRequest.(de?)serializeReplicas

belliottsmith committed:
ReplicaLayout: rename extend->forUncontacted
ReplayLayout nits
restore allReplicas / naturalReplicas usage
do not recompute allReplicas; set all=natural if pending==null

belliottsmith committed:
fix ABWRH.maybeTryAdditionalReplicas for LOCAL_QUORUM

aweisberg commented on src/java/org/apache/cassandra/batchlog/BatchlogManager.java:
```diff
@@ -457,6 +458,7 @@ private void writeHintsForUndeliveredEndpoints(int startFrom, Set<InetAddressAnd
             Keyspace keyspace = Keyspace.open(ks);
             Token tk = mutation.key().getToken();
 
+            // TODO: this should be done through adding "non-avlie" replicas to replica layout
```
Adds a TODO and a spelling error can we resolve this?

belliottsmith replied:
It's not been logically resolved, but there's no need to keep this TODO there specifically so I've removed it - it seems to document the follow-ups to ReplicaLayout Alex and I discussed, that I would very much like to see make it to the codebase, but will have to happen post-4.0

aweisberg commented on src/java/org/apache/cassandra/locator/Endpoints.java:
```diff
@@ -88,14 +78,7 @@ public boolean contains(Replica replica)
     // TODO: TR-Review ignoreConflicts is not an acceptable solution here - we need to explicitly resolve them in case of transient/full mismatch
```
Is this something we can fix now?

belliottsmith replied:
Yes, I'll make a note that we resolved non-isolated mismatches from range movements randomly

aweisberg commented on src/java/org/apache/cassandra/locator/ReplicaLayout.java:
```diff
@@ -0,0 +1,357 @@
+import org.apache.cassandra.service.reads.SpeculativeRetryPolicy;
+import org.apache.cassandra.utils.FBUtilities;
+
+public abstract class ReplicaLayout<E extends Endpoints<E>, L extends ReplicaLayout<E, L>>
```
Needs a class level comment explain what is, what problem it solves, how it came to be etc.

aweisberg commented on src/java/org/apache/cassandra/locator/ReplicaLayout.java:
```diff
@@ -0,0 +1,357 @@
+        this.natural = natural;
+        this.pending = pending;
+        this.selected = selected;
+        if (all == null && pending == null)
```
Need comment?

aweisberg commented on src/java/org/apache/cassandra/locator/ReplicaLayout.java:
```diff
@@ -0,0 +1,357 @@
+    protected final Keyspace keyspace;
+    protected final ConsistencyLevel consistencyLevel;
+
+    private ReplicaLayout(Keyspace keyspace, ConsistencyLevel consistencyLevel, E natural, E pending, E selected)
```
Selected is a new concept to me to some extent so somewhere either the member or the parameter should have a comment explaining what is in it.

belliottsmith replied:
targetReplicas was renamed to selectedReplicas in mainline, so this has simply made the terminology consistent.  A comment would not go amiss, since we now have it in one place, though the class-level comment can probably elaborate all of the semantics.

belliottsmith committed:
Handle transient/full replica conflicts

Commit:
Fix replica selection for non TR case
Fix CL passing for replays
Fix CL calculation for batchlog (as previously it was picked as ONE implicitly through WriteResponseHandler)
Avoid passing unnecessary CL
Put replica selection in “one” place
Use filter instead of select
nit: don't use Iterables.size() on ReplicaCollection (no fast path)
rename {natural,pending,selected,all}Replicas -> {natural,pending,selected,all}
remove usage of intended-to-be-private ReplicaCollection.snapshot()

belliottsmith committed:
fix race condition with hasTransientResponse
cleanup use of Replicas.keep/withoutSelf etc, and move to Endpoints instance method
ensure we use a consistent set of response messages

belliottsmith committed:
fix
